### ISSUE TITLE
What is the right way to specify the Enzyme mode?

### ISSUE THREAD

**Issue by gdalle** [SciML/ADTypes.jl#24]
**Question❓**

The source file says nothing, the test file says something but does something else:
```julia
# In practice, you would rather specify a
# `mode::Enzyme.Mode`, e.g. `Enzyme.Reverse` or `Enzyme.Forward`
adtype = AutoEnzyme(; mode = Val(:Reverse))
@test adtype isa ADTypes.AbstractADType
@test adtype isa AutoEnzyme{Val{:Reverse}}
```
What convention is used in the SciML ecosystem? Can it be documented?

**Comment by Vaibhavdixit02** [comment]
I haven't been using the `mode` kwarg, idk if @avik-pal uses it. For me in Optimization.jl it is just reverse mode for gradient and forward over reverse for hessian so not well suited for the mode argument.

**Comment by avik-pal** [comment]
Me neither I kind of always assume it's reverse mode (which I probably shouldn't)

**Comment by gdalle** [comment]
Another question is what to put inside the AutoEnzyme object, beyond the "mode". Do we need more information to make informed decisions?

**Comment by Vaibhavdixit02** [comment]
>Since you're maintaining packages that make use of ADTypes.jl, I'd be glad to have your opinion on https://github.com/gdalle/DifferentiationInterface.jl.

It looks great! Some other features already being discussed in https://github.com/gdalle/DifferentiationInterface.jl/issues/26, caching etc and support for sparsity will also be pretty essential for me to consider using it in Optimization.jl

**Comment by gdalle** [comment]
@avik-pal There's also another issue with using Enzyme.Reverse as mode, which is that depending on the function (like "value_and_gradient") you may want Enzyme.ReverseWithPrimal instead

**Comment by gdalle** [comment]
@Vaibhavdixit02 indeed! Cache support is fully there for ForwardDiff and ReverseDiff (with and without compiled tape) already, with a unified interface. FiniteDiff and SparseDiffTools are next on my hit list

**Comment by avik-pal** [comment]
> There's also another issue with using Enzyme.Reverse as mode, which is that depending on the function (like "value_and_gradient") you may want Enzyme.ReverseWithPrimal instead

Agreed
